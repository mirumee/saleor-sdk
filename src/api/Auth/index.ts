import { User } from "../../fragments/gqlTypes/User";
import { ErrorListener } from "../../helpers";
import { JobsManager } from "../../jobs";
import { SaleorState, SaleorStateLoaded } from "../../state";
import { StateItems } from "../../state/types";

import { PromiseRunResponse } from "../types";
import { DataErrorAuthTypes, FunctionErrorAuthTypes } from "./types";

export const BROWSER_NO_CREDENTIAL_API_MESSAGE =
  "Saleor SDK is unable to use browser Credential Management API.";

export class AuthAPI extends ErrorListener {
  /**
   * Indicates if data is initialized, initially retrieved from cache or initially fetched.
   */
  loaded: boolean;
  /**
   * User object with currently signed in user data.
   */
  user?: User | null;
  /**
   * Indicates if user is signed in.
   */
  authenticated?: boolean;
  /**
   * Token used for user authentication.
   */
  token?: string;

  private saleorState: SaleorState;
  private jobsManager: JobsManager;

  constructor(saleorState: SaleorState, jobsManager: JobsManager) {
    super();
    this.saleorState = saleorState;
    this.jobsManager = jobsManager;

    this.loaded = false;

    this.saleorState.subscribeToChange(StateItems.USER, (user: User | null) => {
      this.user = user;
      if (this.loaded) {
        this.authenticated = !!this.user;
      }
    });
    this.saleorState.subscribeToChange(StateItems.SIGN_IN_TOKEN, (token) => {
      this.token = token;
    });
    this.saleorState.subscribeToChange(
      StateItems.LOADED,
      (loaded: SaleorStateLoaded) => {
        this.loaded = loaded.user && loaded.signInToken;
        if (this.loaded) {
          this.authenticated = !!this.user;
        }
      }
    );

    if (!this.saleorState.signInToken && window.PasswordCredential) {
      this.autoSignIn();
    }
  }

  /**
   * Tries to authenticate user with given email and password.
   * @param email Email used for authentication.
   * @param password Password used for authentication.
   * @param autoSignIn Indicates if SDK should try to sign in user with given credentials in future without explicitly calling this method. True by default.
   */
  signIn = async (
    email: string,
    password: string,
    autoSignIn: boolean = true
  ): PromiseRunResponse<DataErrorAuthTypes, FunctionErrorAuthTypes> => {
    const { data, dataError } = await this.jobsManager.run("auth", "signIn", {
      email,
      password,
    });

    try {
      if (autoSignIn && !dataError?.error && window.PasswordCredential) {
        await navigator.credentials.store(
          new window.PasswordCredential({
            id: email,
            password,
          })
        );
      }
    } catch (credentialsError) {
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    await this.jobsManager.run("checkout", "provideCheckout", {
      isUserSignedIn: !!data?.user,
    });

    return {
      data,
      dataError,
      pending: false,
    };
  };

  /**
   * Sign out user by clearing cache, local storage and authentication token.
   */
  signOut = async (): PromiseRunResponse<
    DataErrorAuthTypes,
    FunctionErrorAuthTypes
  > => {
    await this.jobsManager.run("auth", "signOut", undefined);
    try {
      if (navigator.credentials && navigator.credentials.preventSilentAccess) {
        await navigator.credentials.preventSilentAccess();
      }
    } catch (credentialsError) {
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    return {
      pending: false,
    };
  };

  private autoSignIn = async () => {
    let credentials;
    try {
      credentials = await (navigator.credentials as any).get({
        password: true,
      });
    } catch (credentialsError) {
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    if (credentials) {
      const { dataError } = await this.signIn(
        credentials.id,
        credentials.password,
        true
      );

      if (dataError?.error) {
        this.fireError(dataError.error, DataErrorAuthTypes.SIGN_IN);
      }
    }
  };
}
