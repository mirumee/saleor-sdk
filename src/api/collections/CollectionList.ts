import BaseList, { GetPageInfo, MapQueryData } from "../../helpers/BaseList";
import { CollectionList as CollectionListQuery } from "../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../fragments/gqlTypes/BaseCollection";

export class CollectionList extends BaseList<
  CollectionListQuery,
  BaseCollection
> {
  getPageInfo: GetPageInfo<CollectionListQuery> = result =>
    result.data.collections?.pageInfo!;

  mapQueryData: MapQueryData<CollectionListQuery, BaseCollection> = data =>
    data.collections?.edges.map(({ node }) => node);
}
