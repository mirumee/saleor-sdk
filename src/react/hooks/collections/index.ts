export { default as useCollectionList } from "./useCollectionList";
