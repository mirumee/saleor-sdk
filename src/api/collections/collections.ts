import ApolloClient from "apollo-client";
import {
  CollectionList as CollectionListQuery,
  CollectionListVariables,
} from "../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../fragments/gqlTypes/BaseCollection";
import { WithList } from "../types";
import { CollectionList } from "./CollectionList";

export class CollectionsAPI
  implements
    WithList<CollectionListQuery, BaseCollection, CollectionListVariables> {
  client: ApolloClient<any>;

  constructor(client: ApolloClient<any>) {
    this.client = client;
  }

  /**
   * Method returning list of collections with ability to request next page
   * @param params List parameters
   */
  getList = (variables: CollectionListVariables): CollectionList => {
    const list = new CollectionList(this.client);

    list.init(variables);

    return list;
  };
}
