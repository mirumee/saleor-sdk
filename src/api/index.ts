import ApolloClient from "apollo-client";

import { defaultConfig } from "../config";
import { LocalStorageManager } from "../data";
import { ApolloClientManager } from "../data/ApolloClientManager";
import { LocalStorageHandler } from "../helpers/LocalStorageHandler";
import { JobsManager } from "../jobs";
import { SaleorState } from "../state";
import { ConfigInput } from "../types";
import { AuthAPI } from "./Auth";
import APIProxy from "./APIProxy";
import { SaleorCartAPI } from "./Cart";
import { SaleorCheckoutAPI } from "./Checkout";
import { CollectionsAPI } from "./collections/collections";

export * from "./Checkout";
export * from "./Cart";

export class SaleorAPI {
  auth: AuthAPI;

  checkout: SaleorCheckoutAPI;

  cart: SaleorCartAPI;

  collections: CollectionsAPI;

  /**
   * @deprecated Please do not use it anymore. Reference to API Proxy will be removed in future.
   * Now it just exists for legacy React hooks, which also will be removed.
   */
  legacyAPIProxy: APIProxy;

  constructor(
    client: ApolloClient<any>,
    apiProxy: APIProxy,
    config: ConfigInput,
    onStateUpdate?: () => any
  ) {
    this.legacyAPIProxy = apiProxy;
    const finalConfig = {
      ...defaultConfig,
      ...config,
      loadOnStart: {
        ...defaultConfig.loadOnStart,
        ...config?.loadOnStart,
      },
    };

    const localStorageHandler = new LocalStorageHandler();
    const apolloClientManager = new ApolloClientManager(client);
    const jobsManager = new JobsManager(
      localStorageHandler,
      apolloClientManager
    );
    const saleorState = new SaleorState(
      finalConfig,
      localStorageHandler,
      apolloClientManager,
      jobsManager
    );
    const localStorageManager = new LocalStorageManager(
      localStorageHandler,
      saleorState
    );

    if (onStateUpdate) {
      saleorState.subscribeToNotifiedChanges(onStateUpdate);
    }

    this.auth = new AuthAPI(saleorState, jobsManager, finalConfig);
    this.checkout = new SaleorCheckoutAPI(saleorState, jobsManager);
    this.cart = new SaleorCartAPI(
      localStorageManager,
      apolloClientManager,
      saleorState,
      jobsManager
    );
    this.collections = new CollectionsAPI(client);

    this.legacyAPIProxy.attachAuthListener(authenticated => {
      if (!authenticated) {
        localStorageHandler.setCheckout({});
        localStorageHandler.setPayment({});
        localStorageHandler.setJobs(null);
      }
    });
  }
}
