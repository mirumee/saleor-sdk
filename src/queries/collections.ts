import gql from "graphql-tag";
import { baseCollectionFragment } from "../fragments/collections";
import { pageInfo } from "../fragments/pageInfo";

export const collections = gql`
  ${baseCollectionFragment}
  ${pageInfo}
  query CollectionList(
    $first: Int!
    $after: String
    $sortBy: CollectionSortingInput
    $filter: CollectionFilterInput
  ) {
    collections(
      first: $first
      after: $after
      sortBy: $sortBy
      filter: $filter
    ) {
      edges {
        node {
          ...BaseCollection
        }
      }
      pageInfo {
        ...PageInfo
      }
    }
  }
`;
