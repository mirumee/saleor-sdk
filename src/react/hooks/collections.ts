import { CollectionList } from "../../api/collections/CollectionList";
import { makeList } from "./utils";

export const useCollectionList = makeList(client => new CollectionList(client));
