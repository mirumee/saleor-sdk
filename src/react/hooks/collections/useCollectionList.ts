import { useApolloClient } from "react-apollo";
import { useEffect, useRef, useState } from "react";
import ApolloClient from "apollo-client";
import {
  CollectionList as CollectionListQuery,
  CollectionListVariables,
} from "../../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../../fragments/gqlTypes/BaseCollection";
import { ListParameters } from "../../../api/types";
import { collections } from "../../../queries/collections";
import { CollectionList } from "../../../api/collections/CollectionList";

function query(client: ApolloClient<any>) {
  return (variables: CollectionListVariables) =>
    client.query<CollectionListQuery, CollectionListVariables>({
      query: collections,
      variables,
    });
}

function useCollectionList(params: ListParameters) {
  const client = useApolloClient();
  const [data, setData] = useState<BaseCollection[] | undefined>(undefined);
  const list = useRef<CollectionList>();
  const willMount = useRef(true);

  async function initList() {
    list.current = new CollectionList(query(client), params.count);
    await list.current.init({
      after: params.endCursor,
      first: params.count,
    });
    setData(list.current.data);
  }

  useEffect(() => {
    if (!willMount.current) {
      initList();
    }
  }, [params.count, params.endCursor]);

  if (willMount.current) {
    initList();
    willMount.current = false;
  }

  const next = async () => {
    const nextData = await list.current?.next();
    setData(nextData);
  };

  return {
    current: list.current?.current,
    data,
    loading: list.current?.loading,
    next,
    pageInfo: list.current?.pageInfo,
  };
}

export default useCollectionList;
