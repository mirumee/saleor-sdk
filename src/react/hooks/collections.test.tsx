import { act, renderHook } from "@testing-library/react-hooks";
import React from "react";
import { ApolloProvider } from "react-apollo";
import { setupRecording, setupAPI } from "../../../testUtils/api";
import { useCollectionList } from "./collections";
import {
  OrderDirection,
  CollectionSortField,
} from "../../gqlTypes/globalTypes";

setupRecording();

describe("useCollectionList", () => {
  let wrapper: React.FC<{}>;

  beforeAll(async () => {
    const { client } = await setupAPI();

    wrapper = ({ children }) => (
      <ApolloProvider client={client}>{children}</ApolloProvider>
    );
  });

  it("can fetch collections", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          first: 20,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can fetch next page", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          first: 1,
        }),
      {
        wrapper,
      }
    );

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();

    await act(result.current.next);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can sort", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          first: 10,
          sortBy: {
            direction: OrderDirection.DESC,
            field: CollectionSortField.NAME,
          },
        }),
      {
        wrapper,
      }
    );

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
  });

  it("can filter", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          filter: {
            search: "winter",
          },
          first: 10,
        }),
      {
        wrapper,
      }
    );

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
  });
});
