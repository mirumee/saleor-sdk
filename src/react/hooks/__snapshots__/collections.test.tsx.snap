// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCollectionList can fetch collections 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;

exports[`useCollectionList can fetch next page 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
]
`;

exports[`useCollectionList can fetch next page 2`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;

exports[`useCollectionList can filter 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;

exports[`useCollectionList can sort 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
]
`;
